### DISCUSSION TITLE
during collection.load(), why is there only a small amount of query nodes doing work?

### DISCUSSION THREAD

**Discussion by aping** [milvus-io/milvus#33238] [Q&A and General discussion] [answered]
I ingested ~250M vectors into a collection, with 1000 partitions. and I started 6 query nodes, each one with 256GB of ram.
problem is, when I do `collection.load()`, I can see that only 1 query node was doing job, the rest 5 are just idle.
then the one that does actual work eventually OOM, and then another query node picks up the job, and OOMs eventually.
so I'm wondering there must be something that I'm missing? why is only 1/6 query nodes doing work?

PS: the server version is 2.4.1
PS2: I also tried 8x 128GB query nodes, in that case 2-3 query nodes were doing actual work and ooms too eventually

**Comment by aping** [comment]
this looks bad?
```
show segment
...
SegmentID: 449872538076379937 State: Flushed, Row Count:83746
--- Growing: 6, Sealed: 0, Flushed: 21248
--- Total Segments: 21254, row count: 265462841
```
21k segments, that’s like 12k rows per segment, the dim is 768 so each segment is small.
so questions are:
- does this have anything to do with only 1/6 query nodes are loading stuff into memory, the rest are sitting idle?
- should I call `flush()`?
- or it is better if I call `compact()` since the segments are very small?

**Reply by xiaofan-luan** [reply]
did you build index done?

please offer your querycoord and querynode log so we can know the reason.
Possible reason:
1. index building is not done, causing all data can only be loaded on delegator
2. you enabled some feature like resource group, and all the collection is load to same resource group

**Reply by aping** [reply]
> did you build index done?

`get_index_build_progress` returns
```
{'total_rows': 265368892, 'indexed_rows': 265368892, 'pending_index_rows': 0}
```
so I assume index building is done?

> you enabled some feature like resource group
```
for rg in connection.list_resource_groups():
    detail = connection.describe_resource_group(rg)
    print(f"resource group {rg}: {detail}")
        
resource group __default_resource_group: ResourceGroupInfo:
<name:__default_resource_group>,
<capacity:1000000>,
<num_available_node:6>,
...
```
so again i don't think so? just the default resource group.
[logs.zip](https://github.com/milvus-io/milvus/files/15397323/logs.zip)

**Comment by xiaofan-luan** [comment]
i think you have too many partitions + shards. 

we will try to improve on this. but brings too heavy workload and is not recommended

**Comment by xiaofan-luan** [comment]
all the data adding is to load growing segment and seems growing segment amount is huge

![Uploading image.png…]()

**Comment by xiaofan-luan** [comment]
i think the problem is 449797948377841297. 
this colleciton has more than 7000+ growing segments

**Comment by xiaofan-luan** [comment]
and if you take a look at SegmentID: 449872538076379937 State: Flushed, Row Count:83746
--- Growing: 6, Sealed: 0, Flushed: 21248
--- Total Segments: 21254, row count: 265462841

with only 83746 rows, you have more than 21248 segments. which is definitely not milvus designed for

**Reply by xiaofan-luan** [reply]
please remove the 1000 partition design and retry.

**Comment by xiaofan-luan** [comment]
if you can help us to get a pprof on the busy querynode that could be really helpful as well.

**Comment by xiaofan-luan** [comment]
from your use case, gc is super slow 

[2024/05/22 02:22:02.074 +00:00] [WARN] [gc/gc_tuner.go:99] ["GC Tune done, and the gc is slow"] ["previous GOGC"=30] ["heapuse "=169709] ["total memory"=202095] ["next GC"=191921] ["new GOGC"=30] [gc-pause=1.228162124s] [gc-pause-end=1716344521546144595]

and heap use is huge. which means something in go takes a lot of memory but we need pprof to see what it is

**Comment by aping** [comment]
yes there's only one collection, with 1000 partitions. I thought with 250M rows, there's still like 250k rows per partition, that's like `250k * 768 (dimension) * 4 (bytes per float) = 730MB` per partition, not that bad?
as for pprof, I'm using helm and the official image `milvusdb/milvus:v2.4.1`, not sure how to enable pprof?

a few follow up questions:
- I understant that I have too many small/fragmented segment, but does this have anything to do with only one query node is working during `collection.load` and the rest are idle?
- will `collection.compact` help?
- I created the collection with 1000 partitions because I expect it will have 5-10B rows eventually. so if I create it with say 100 partitions, and when the data grows, I don't think I can increase the partition count, when I have partition key enabled?

**Reply by xiaofan-luan** [reply]
we already enabeld pprof. you should be able to connect at http://ip:9091/debug/pprof/

and profile and heap could be helpful.


1. We see at least collections in your environement,  449872538076379937 and 449797948377841297, please check why.
2. 449797948377841297 has many growing segments, that's the major reason load can not be succesful. 
3. compact won't help on this case, because shard and partition is huge.
4. Did you search by partition? or the whole collection. If search partition, then 1000 partition might be reasonable under 5-10B data.

We'd like learn more about your use cases and help you， I don't think it's a easy task to handle 10B vector

**Reply by aping** [reply]
1: 449872538076379937 is useless, a very small one, I don't even touch it
2-3: so help me understand, I have 4 shards, 1000 partitions, what is the least number of segments, in a perfect situation? is it one per partition, so that's like at least 1000, or one per partition per shard, that's like 4000?
4: yes partition key will be used during searching. I can easily create another col with 100 partitions and upload data, but again, if my data grows to 5B rows, what would be my option then?

and finally, again, any idea why only one query node was actually loading data during `collection.load`? I mean even with the absurd number of segments, with 6x128GB query nodes, if they can balance, at least I should be able to load them instead of OOM

**Reply by xiaofan-luan** [reply]
we don't know what is the segment number of 449797948377841297, but from current log growing segment is already too much(7000 + )

the problem is your index build and datanode flush can not catch up and all the segment is not convert to sealed.

**Reply by xiaofan-luan** [reply]
with many partitions, you have to tune performance and do quota limitation very carefully and this is not recommend for a starter of milvus. 

With that much data, I'm assuming your QPS will be very low(less than 10), create more partitions can help on performance but you will need to carefully balance resource of datanode, indexnode so i don't recommend this

**Reply by aping** [reply]
> we don't know what is the segment number of 449797948377841297, but from current log growing segment is already too much(7000 + )

how can you tell there're 7000+ growing segments, I mean birdwatch says
```
show segments
...
SegmentID: 449797948406082581 State: Flushed, Row Count:87172
SegmentID: 449872538076379937 State: Flushed, Row Count:83746
--- Growing: 0, Sealed: 0, Flushed: 21254
--- Total Segments: 21254, row count: 265462841
```

> the problem is your index build and datanode flush can not catch up and all the segment is not convert to sealed.

how do I check if index nodes and data nodes are not catching up? I mean I see almost 0 activity with datanode and index node if I don't do `collection.load` ? and `get_index_build_progress` returns
```
{'total_rows': 265368892, 'indexed_rows': 265368892, 'pending_index_rows': 0}
```
so I thought the writing part is done, except that the segments are too fragmented?

> With that much data, I'm assuming your QPS will be very low(less than 10), create more partitions can help on performance but you will need to carefully balance resource of datanode, indexnode so i don't recommend this

yes it's for internal use so I expect qps to be low but the data size can be huge, only thing stopping us is the hardware cost.

> We'd like learn more about your use cases and help you

feel free to email me, I'm not sure if it's ok for me to discuss this publicly

and finally results from `curl http://localhost:9091/debug/pprof/profile?seconds=10 > profile.out` and `http://localhost:9091/debug/pprof/heap > heap.out`
[pprof.zip](https://github.com/milvus-io/milvus/files/15400531/pprof.zip)

**Comment by xiaofan-luan** [comment]
> > we don't know what is the segment number of 449797948377841297, but from current log growing segment is already too much(7000 + )
> 
> how can you tell there're 7000+ growing segments, I mean birdwatch says
> 
> ```
> show segments
> ...
> SegmentID: 449797948406082581 State: Flushed, Row Count:87172
> SegmentID: 449872538076379937 State: Flushed, Row Count:83746
> --- Growing: 0, Sealed: 0, Flushed: 21254
> --- Total Segments: 21254, row count: 265462841
> ```
> 
> > the problem is your index build and datanode flush can not catch up and all the segment is not convert to sealed.
> 
> how do I check if index nodes and data nodes are not catching up? I mean I see almost 0 activity with datanode and index node if I don't do `collection.load` ? and `get_index_build_progress` returns
> 
> ```
> {'total_rows': 265368892, 'indexed_rows': 265368892, 'pending_index_rows': 0}
> ```
> 
> so I thought the writing part is done, except that the segments are too fragmented?
> 
> > With that much data, I'm assuming your QPS will be very low(less than 10), create more partitions can help on performance but you will need to carefully balance resource of datanode, indexnode so i don't recommend this
> 
> yes it's for internal use so I expect qps to be low but the data size can be huge, only thing stopping us is the hardware cost.
> 
> > We'd like learn more about your use cases and help you
> 
> feel free to email me, I'm not sure if it's ok for me to discuss this publicly
> 
> and finally results from `curl http://localhost:9091/debug/pprof/profile?seconds=10 > profile.out` and `http://localhost:9091/debug/pprof/heap > heap.out` [pprof.zip](https://github.com/milvus-io/milvus/files/15400531/pprof.zip)

what you show is the smaller colleciton, larger segment should have larger number of segment.
please contact me at xiaofan.luan@zilliz.com

I will follow this pprof

**Comment by xiaofan-luan** [comment]
too many deletes

<img width="1008" alt="image" src="https://github.com/milvus-io/milvus/assets/83447078/e0e9efe2-8c4b-4b9a-89b0-6354d2fbac97">

**Reply by aping** [reply]
what does Delete mean here? I do use `upsert` instead of `insert`, but the collection was only `upsert` once so there should not be any duplicates

**Reply by xiaofan-luan** [reply]
upsert will generate many unnessary deletes

**Reply by aping** [reply]
so a few questions:
- how does upsert work? say for example I upsert 1000 rows, does it first generate 1000 deletes (does not check if the pks exist), then generate 1000 inserts?
- would compaction merge these records eventually?
- if I want to batch insert say 1B rows in 24h, should I avoid upsert at all?

**Comment by xiaofan-luan** [comment]
> 1. 449797948377841297



> so a few questions:
> 
> * how does upsert work? say for example I upsert 1000 rows, does it first generate 1000 deletes (does not check if the pks exist), then generate 1000 inserts?
> * would compaction merge these records eventually?
> * if I want to batch insert say 1B rows in 24h, should I avoid upsert at all?

1. current implementation might generate delete far more than that. when you have large number of segment(10000), even if each segment has relatively low false positive rate(0.05), you still generate 500 delete  on each upsert, that need to be improved.
2. yes, but compaction will take a very long time to do this, you can tune compaction parameter and this is for sure can be optimized
3. yes, we don't expect upsert to happen that often, especially on such large dataset.

**Reply by aping** [reply, marked as answer]
ok reuploaded with insert, everything worked find.
- 340M rows, 768 dim
- upload/insert took 3h
- indexing took ~14h with 4x8c index nodes
- load took 26m, total of 350GB memory used by query nodes
- search latency is totally acceptable, <1s with 10 qps
I was originally hoping to use upsert to avoid data duplicates, looks like it does not work like I had expected, but I think I can do something like
```
def upload(data):
  exists = collection.get(data.ids)
  if len(exists) > 0:
    collection.delete(expr="id in [exists...]")
  collection.insert(data)
```
?

**Reply by xiaofan-luan** [reply]
current unfrequent delete is ok. 
we are still improving performance with many deletes
